pkg/clusteragent/autoscaling/workload/model/pod_autoscaler_test.go

```diff
@@ -0,0 +1,78 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver && test
+
+package model
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	datadoghq "github.com/DataDog/datadog-operator/apis/datadoghq/v1alpha1"
+)
+
+func TestAddHorizontalAction(t *testing.T) {
+	testTime := time.Now()
+
+	// Test no retention, should move back to keep a single action
+	var horizontalEventsRetention time.Duration
+	horizontalLastActions := []datadoghq.DatadogPodAutoscalerHorizontalAction{
+		{
+			Time: metav1.Time{Time: testTime.Add(-10 * time.Minute)},
+		},
+		{
+			Time: metav1.Time{Time: testTime.Add(-8 * time.Minute)},
+		},
+	}
+	addedAction1 := &datadoghq.DatadogPodAutoscalerHorizontalAction{
+		Time: metav1.Time{Time: testTime},
+	}
+	horizontalLastActions = addHorizontalAction(testTime, horizontalEventsRetention, horizontalLastActions, addedAction1)
+	assert.Equal(t, []datadoghq.DatadogPodAutoscalerHorizontalAction{*addedAction1}, horizontalLastActions)
+
+	// Add another event, should still keep one
+	horizontalLastActions = addHorizontalAction(testTime, horizontalEventsRetention, horizontalLastActions, addedAction1)
+	assert.Equal(t, []datadoghq.DatadogPodAutoscalerHorizontalAction{*addedAction1}, horizontalLastActions)
+
+	// 15 minutes retention, should keep everything
+	horizontalEventsRetention = 15 * time.Minute
+	horizontalLastActions = []datadoghq.DatadogPodAutoscalerHorizontalAction{
+		{
+			Time: metav1.Time{Time: testTime.Add(-10 * time.Minute)},
+		},
+		{
+			Time: metav1.Time{Time: testTime.Add(-8 * time.Minute)},
+		},
+	}
+	// Adding two fake events
+	horizontalLastActions = addHorizontalAction(testTime, horizontalEventsRetention, horizontalLastActions, addedAction1)
+	horizontalLastActions = addHorizontalAction(testTime, horizontalEventsRetention, horizontalLastActions, addedAction1)
+	assert.Equal(t, []datadoghq.DatadogPodAutoscalerHorizontalAction{
+		{
+			Time: metav1.Time{Time: testTime.Add(-10 * time.Minute)},
+		},
+		{
+			Time: metav1.Time{Time: testTime.Add(-8 * time.Minute)},
+		},
+		*addedAction1,
+		*addedAction1,
+	}, horizontalLastActions)
+
+	// Moving time forward, should keep only the last two events
+	testTime = testTime.Add(10 * time.Minute)
+	addedAction2 := &datadoghq.DatadogPodAutoscalerHorizontalAction{
+		Time: metav1.Time{Time: testTime},
+	}
+	horizontalLastActions = addHorizontalAction(testTime, horizontalEventsRetention, horizontalLastActions, addedAction2)
+	assert.Equal(t, []datadoghq.DatadogPodAutoscalerHorizontalAction{
+		*addedAction1,
+		*addedAction1,
+		*addedAction2,
+	}, horizontalLastActions)
+}
```

Comment:
```suggestion
func TestAddHorizontalAction(t *testing.T) {
	testTime := time.Now()
	actionNow := &datadoghq.DatadogPodAutoscalerHorizontalAction{
		Time: metav1.Time{Time: testTime},
	}
	actionIn10Min := &datadoghq.DatadogPodAutoscalerHorizontalAction{
		Time: metav1.Time{Time: testTime.Add(10 * time.Minute)},
	}

	testCases := []struct {
		description               string
		testTime                  time.Time
		horizontalEventsRetention time.Duration
		horizontalLastActions     []datadoghq.DatadogPodAutoscalerHorizontalAction
		addedAction               *datadoghq.DatadogPodAutoscalerHorizontalAction
		expectedActions           []datadoghq.DatadogPodAutoscalerHorizontalAction
	}{
		{
			description: "No retention, keep single action",
			testTime:    testTime,
			horizontalLastActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{
				{Time: metav1.Time{Time: testTime.Add(-10 * time.Minute)}},
				{Time: metav1.Time{Time: testTime.Add(-8 * time.Minute)}},
			},
			addedAction:     actionNow,
			expectedActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{*actionNow},
		},
		{
			description: "Add another event, still keep one",
			testTime:    testTime,
			horizontalLastActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{
				*actionNow,
			},
			addedAction:     actionNow,
			expectedActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{*actionNow},
		},
		{
			description:               "15 minutes retention, keep everything",
			testTime:                  testTime,
			horizontalEventsRetention: 15 * time.Minute,
			horizontalLastActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{
				{Time: metav1.Time{Time: testTime.Add(-10 * time.Minute)}},
				{Time: metav1.Time{Time: testTime.Add(-8 * time.Minute)}},
			},
			addedAction: actionNow,
			expectedActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{
				{Time: metav1.Time{Time: testTime.Add(-10 * time.Minute)}},
				{Time: metav1.Time{Time: testTime.Add(-8 * time.Minute)}},
				*actionNow,
			},
		},
		{
			description:               "Moving time forward, keep only last two events",
			testTime:                  testTime.Add(10 * time.Minute),
			horizontalEventsRetention: 15 * time.Minute,
			horizontalLastActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{
				*actionNow,
				*actionNow,
			},
			addedAction: actionIn10Min,
			expectedActions: []datadoghq.DatadogPodAutoscalerHorizontalAction{
				*actionNow,
				*actionNow,
				*actionIn10Min,
			},
		},
	}

	for _, tc := range testCases {
		t.Run(tc.description, func(t *testing.T) {
			result := addHorizontalAction(tc.testTime, tc.horizontalEventsRetention, tc.horizontalLastActions, tc.addedAction)
			assert.Equal(t, tc.expectedActions, result)
		})
	}
}
```

Reply:
I am not a big fan of array-based testing in this case as the scenario is sequential and the sequence order is important.